File: src/main/java/org/radarcns/management/service/mapper/SourceDataMapper.java
package org.radarcns.management.service.mapper;

import org.mapstruct.IterableMapping;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Named;
import org.radarcns.management.domain.SourceData;
import org.radarcns.management.service.dto.SourceDataDTO;

import java.util.List;

/**
 * Mapper for the entity SourceData and its DTO SourceDataDTO.
 */
@Mapper(componentModel = "spring", uses = {SourceTypeMapper.class})
public interface SourceDataMapper {

    @Named("sourceDataDTO")
    SourceDataDTO sourceDataToSourceDataDTO(SourceData sourceData);

    @Named("sourceDataReducedDTO")
    @Mapping(target = "sourceType", ignore = true)
    SourceDataDTO sourceDataToSourceDataDTOReduced(SourceData sourceData);

    @IterableMapping(qualifiedByName = "sourceDataReducedDTO")
    List<SourceDataDTO> sourceDataToSourceDataDTOs(List<SourceData> sourceData);

    SourceData sourceDataDTOToSourceData(SourceDataDTO sourceDataDto);

    List<SourceData> sourceDataDTOsToSourceData(List<SourceDataDTO> sourceDataDtos);

    /**
     * generating the fromId for all mappers if the databaseType is sql, as the class has
     * relationship to it might need it, instead of creating a new attribute to know if the entity
     * has any relationship from some other entity.
     *
     * @param id id of the entity
     * @return the entity instance
     */

    default SourceData sourceDataFromId(Long id) {
        if (id == null) {
            return null;
        }
        SourceData sourceData = new SourceData();
        sourceData.setId(id);
        return sourceData;
    }
}

File: src/main/java/org/radarbase/management/service/mapper/SourceDataMapper.kt
package org.radarbase.management.service.mapper

import org.mapstruct.IterableMapping
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.Named
import org.radarbase.management.domain.SourceData
import org.radarbase.management.service.dto.SourceDataDTO

/**
 * Mapper for the entity SourceData and its DTO SourceDataDTO.
 */
@Mapper(componentModel = "spring", uses = [SourceTypeMapper::class])
interface SourceDataMapper {
    @Named("sourceDataDTO")
    fun sourceDataToSourceDataDTO(sourceData: SourceData?): SourceDataDTO?

    @Named("sourceDataReducedDTO")
    @Mapping(target = "sourceType", ignore = true)
    fun sourceDataToSourceDataDTOReduced(sourceData: SourceData): SourceDataDTO

    @IterableMapping(qualifiedByName = ["sourceDataReducedDTO"])
    fun sourceDataToSourceDataDTOs(sourceData: List<SourceData>): List<SourceDataDTO>
    fun sourceDataDTOToSourceData(sourceDataDto: SourceDataDTO): SourceData
    fun sourceDataDTOsToSourceData(sourceDataDtos: List<SourceDataDTO>): List<SourceData>

    /**
     * generating the fromId for all mappers if the databaseType is sql, as the class has
     * relationship to it might need it, instead of creating a new attribute to know if the entity
     * has any relationship from some other entity.
     *
     * @param id id of the entity
     * @return the entity instance
     */
    fun sourceDataFromId(id: Long?): SourceData? {
        if (id == null) {
            return null
        }
        val sourceData = SourceData()
        sourceData.id = id
        return sourceData
    }
}
